helios-services/src/main/java/com/spotify/helios/master/resources/HostsResource.java

```diff
@@ -88,8 +93,35 @@ public HostsResource(final MasterModel model) {
   @Produces(APPLICATION_JSON)
   @Timed
   @ExceptionMetered
-  public List<String> list() {
-    return model.listHosts();
+  public List<String> list(@QueryParam("namePattern") final String namePattern,
+                           @QueryParam("selector") final List<String> hostSelectors) {
+
+    List<String> hosts = model.listHosts();
+
+    if (namePattern != null) {
+      final Predicate<String> matchesPattern = Pattern.compile(namePattern).asPredicate();
+      hosts = hosts.stream()
+          .filter(matchesPattern)
+          .collect(Collectors.toList());
+    }
+
+    if (!hostSelectors.isEmpty()) {
+
+      final Map<String, Map<String, String>> hostsAndLabels = hosts.stream()
+          .collect(Collectors.toMap(Function.identity(),
+                                    host -> model.getHostStatus(host).getLabels()
+                   )
+          );
+
+      final List<HostSelector> selectors = hostSelectors.stream()
+          .map(HostSelector::parse)
```

Comment:
note that `HostSelector::parse` returns `null` on bad input

I'm guessing this will yield a 500 with no error message on bad input?